src/redisearch_rs/c_entrypoint/query_error_ffi/src/lib.rs

```diff
@@ -0,0 +1,347 @@
+
+/// Returns the default [`QueryError`].
+///
+/// # Safety
+///
+/// `OpaqueQueryError` and `QueryError` must have the same size and alignment.
+#[unsafe(no_mangle)]
+pub unsafe extern "C" fn QueryError_Default() -> OpaqueQueryError {
+    QueryError::default().into_opaque()
+}
+
+/// Returns true if `query_error` has no error code set.
+///
+/// # Safety
+///
+/// `query_error` must have been created by [`QueryError_Default`].
+#[unsafe(no_mangle)]
+pub unsafe extern "C" fn QueryError_IsOk(query_error: *const OpaqueQueryError) -> bool {
+    // Safety: see safety requirement above.
+    let query_error =
+        unsafe { QueryError::from_opaque_ptr(query_error) }.expect("query_error is null");
+
+    query_error.is_ok()
+}
+
+/// Returns true if `query_error` has an error code set.
+///
+/// # Safety
+///
+/// `query_error` must have been created by [`QueryError_Default`].
+#[unsafe(no_mangle)]
+pub unsafe extern "C" fn QueryError_HasError(query_error: *const OpaqueQueryError) -> bool {
+    // Safety: see safety requirement above.
+    unsafe { !QueryError_IsOk(query_error) }
+}
+
+/// Returns a human-readable string representing the provided [`QueryErrorCode`].
+///
+/// # Safety
+///
+/// This function should always return without a panic for any value provided.
+/// It is unique among the `QueryError_*` API as the only function which allows
+/// an invalid [`QueryErrorCode`] to be provided.
+#[unsafe(no_mangle)]
+pub unsafe extern "C" fn QueryError_Strerror(maybe_code: u8) -> *const c_char {
+    let Ok(code) = QueryErrorCode::try_from(maybe_code) else {
+        return c"Unknown status code".as_ptr();
+    };
+
+    code.to_c_str().as_ptr()
+}
+
+/// Sets the [`QueryErrorCode`] and error message for a [`QueryError`].
+///
+/// This does not mutate `query_error` if it already has an error set.
+///
+/// # Safety
+///
+/// - `query_error` must have been created by [`QueryError_Default`].
+/// - `code` must be a valid variant of [`QueryErrorCode`].
+/// - `message` must be a valid C string or a NULL pointer.
+#[unsafe(no_mangle)]
+pub unsafe extern "C" fn QueryError_SetError(
+    query_error: *mut OpaqueQueryError,
+    code: u8,
+    message: *const c_char,
+) {
+    // Safety: see safety requirement above.
+    let query_error =
+        unsafe { QueryError::from_opaque_mut_ptr(query_error) }.expect("query_error is null");
+    let code = QueryErrorCode::try_from(code).expect("invalid query error code");
+
+    let message = if message.is_null() {
+        None
+    } else {
+        // Safety: see safety requirement above.
+        Some(unsafe { CStr::from_ptr(message) }.to_owned())
+    };
+
+    query_error.set_code_and_message(code, message);
+}
+
+/// Sets the [`QueryErrorCode`] for a [`QueryError`].
+///
+/// This does not mutate `query_error` if it already has an error set.
+///
+/// # Safety
+///
+/// - `query_error` must have been created by [`QueryError_Default`].
+/// - `code` must be a valid variant of [`QueryErrorCode`].
```

Comment:
A panic occurs (line 113) if that code is invalid, not UB. Let's put it under a `# Panics` header instead